ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandRunner.java

```diff
@@ -56,51 +64,55 @@ public CommandRunner(
       final KsqlEngine ksqlEngine,
       final int maxRetries,
       final ClusterTerminator clusterTerminator
+  ) {
+    this(
+        statementExecutor,
+        commandStore,
+        ksqlEngine,
+        maxRetries,
+        clusterTerminator,
+        Executors.newSingleThreadExecutor(r -> new Thread(r, "CommandRunner"))
+    );
+  }
+
+  @VisibleForTesting
+  CommandRunner(
+      final StatementExecutor statementExecutor,
+      final CommandQueue commandStore,
+      final KsqlEngine ksqlEngine,
+      final int maxRetries,
+      final ClusterTerminator clusterTerminator,
+      final ExecutorService executor
   ) {
     this.statementExecutor = Objects.requireNonNull(statementExecutor, "statementExecutor");
     this.commandStore = Objects.requireNonNull(commandStore, "commandStore");
     this.ksqlEngine = Objects.requireNonNull(ksqlEngine, "ksqlEngine");
     this.maxRetries = maxRetries;
-    closed = false;
     this.clusterTerminator = Objects.requireNonNull(clusterTerminator, "clusterTerminator");
+    this.executor = Objects.requireNonNull(executor, "executor");
   }
 
   /**
    * Begin a continuous poll-execute loop for the command topic, stopping only if either a
    * {@link WakeupException} is thrown or the {@link #close()} method is called.
    */
-  @Override
-  public void run() {
-    try {
-      while (!closed) {
-        log.debug("Polling for new writes to command topic");
-        fetchAndRunCommands();
-      }
-    } catch (final WakeupException wue) {
-      if (!closed) {
-        throw wue;
-      }
-    }
+  public void start() {
+    executor.submit(new Runner());
+    executor.shutdown();
   }
 
   /**
    * Halt the poll-execute loop.
    */
   @Override
   public void close() {
-    closed = true;
-    commandStore.close();
-  }
-
-  void fetchAndRunCommands() {
-    final List<QueuedCommand> commands = commandStore.getNewCommands();
-    final Optional<QueuedCommand> terminateCmd = findTerminateCommand(commands);
-    if (terminateCmd.isPresent()) {
-      terminateCluster(terminateCmd.get().getCommand());
-      return;
+    try {
+      closed = true;
+      executor.awaitTermination(SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```

Comment:
It would be nice to do a wakeup of the consumer here before calling awaitTermination. Also, CommandStore.close has an unnecessary wakeup we can clean up.

Reply:
Tis true. Done.